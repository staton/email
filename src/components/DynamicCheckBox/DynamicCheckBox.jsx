import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {MdCheck} from 'react-icons/md';
import BackgroundColor from '../../enums/backgroundColor';

const propTypes = {
    backgroundColor: PropTypes.string,
    content: PropTypes.string,
    isActive: PropTypes.bool,
    isChecked: PropTypes.bool,
    onClick: PropTypes.func
};

const defaultProps = {
    backgroundColor: BackgroundColor.Gray,
    content: '',
    isActive: false,
    isChecked: false,
    onClick: () => {}
};

class DynamicCheckBox extends Component {

    constructor() {
        super();

        this.handleClicked = this.handleClicked.bind(this);
    }

	render() {
		return (
            <div 
                className="DynamicCheckBox"
                onClick={this.handleClicked}
            >
                <div className={this.getContainerClassName()}>
                    <div 
                        className="DynamicCheckBox__letter card__front"
                        style={{backgroundColor: this.props.backgroundColor}}
                    >
                        {this.props.content}
                    </div>
                    <div className={this.getCheckBoxClassName()}>
                        {
                            (this.props.isChecked)
                            ?   <MdCheck />
                            :   null
                        }
                    </div>
                </div>
            </div>
		);
    }

    getCheckBoxClassName() {
        let className = 'DynamicCheckBox__check card__back';

        if (this.props.isChecked) 
            className += ' DynamicCheckBox__check--checked'

        return className;
    }

    /**
     * Gets the class name(s) for the inner container.
     * @returns {string} The class name(s).
     */
    getContainerClassName() {
        let className = 'DynamicCheckBox__container card';

        if (this.props.isActive)
            className += ' card--is-flipped';


        return className;
    }

    /**
     * Called when the checkbox is clicked.
     * @param {object} e The event.
     */
    handleClicked(e) {
        if (this.props.isActive) {
            e.preventDefault();
            e.stopPropagation();
            this.props.onClick();
        }
    }

}

DynamicCheckBox.propTypes = propTypes;
DynamicCheckBox.defaultProps = defaultProps;

export default DynamicCheckBox;