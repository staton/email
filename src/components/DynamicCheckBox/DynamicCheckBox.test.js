import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';
import DynamicCheckBox from './DynamicCheckBox';
import {MdCheck} from 'react-icons/md';

describe('<DynamicCheckBox />', () => {

    it('renders correctly when not active and unchecked', () => {
        const wrapper = shallow(
            <DynamicCheckBox 
                isActive={false}
                isChecked={false} 
            />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders correctly when active and unchecked', () => {
        const wrapper = shallow(
            <DynamicCheckBox 
                isActive={true}
                isChecked={false} 
            />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders correctly when active and checked', () => {
        const wrapper = shallow(
            <DynamicCheckBox 
                isActive={true}
                isChecked={true} 
            />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('has a check icon when checked', () => {
        const wrapper = shallow(
            <DynamicCheckBox 
                isActive={true}
                isChecked={true} 
            />);
        expect(wrapper.find(MdCheck).length).toEqual(1);
        expect(wrapper.find('.DynamicCheckBox__check--checked').length).toEqual(1);
    });

    it('does not have a check icon when unchecked', () => {
        const wrapper = shallow(
            <DynamicCheckBox 
                isActive={true}
                isChecked={false} 
            />);
        expect(wrapper.find(MdCheck).length).toEqual(0);
        expect(wrapper.find('.DynamicCheckBox__check--checked').length).toEqual(0);
    });

    it('has the .card--is-flipped class when active', () => {
        const wrapper = shallow(
            <DynamicCheckBox 
                isActive={true}
            />);
        expect(wrapper.find('.card--is-flipped').length).toEqual(1);
    });

    it('does not have the .card--is-flipped class when not active', () => {
        const wrapper = shallow(
            <DynamicCheckBox 
                isActive={false}
            />);
        expect(wrapper.find('.card--is-flipped').length).toEqual(0);
    });
    
    it('handles click events when active', () => {
        const onClick = jest.fn();
        const wrapper = mount(
            <DynamicCheckBox 
                isActive={true}
                onClick={onClick} 
            />);
        let button = wrapper.find('.DynamicCheckBox');

        button.simulate('click');
        expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('ignores click events when not active', () => {
        const onClick = jest.fn();
        const wrapper = mount(
            <DynamicCheckBox 
                isActive={false}
                onClick={onClick} 
            />);
        let button = wrapper.find('.DynamicCheckBox');

        button.simulate('click');
        expect(onClick).toHaveBeenCalledTimes(0);
    });
    
});