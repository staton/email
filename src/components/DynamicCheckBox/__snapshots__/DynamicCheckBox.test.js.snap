// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicCheckBox /> renders correctly when active and checked 1`] = `
<div
  className="DynamicCheckBox"
  onClick={[Function]}
>
  <div
    className="DynamicCheckBox__container card card--is-flipped"
  >
    <div
      className="DynamicCheckBox__letter card__front"
      style={
        Object {
          "backgroundColor": "rgb(189, 189, 189)",
        }
      }
    />
    <div
      className="DynamicCheckBox__check card__back DynamicCheckBox__check--checked"
    >
      <Component />
    </div>
  </div>
</div>
`;

exports[`<DynamicCheckBox /> renders correctly when active and unchecked 1`] = `
<div
  className="DynamicCheckBox"
  onClick={[Function]}
>
  <div
    className="DynamicCheckBox__container card card--is-flipped"
  >
    <div
      className="DynamicCheckBox__letter card__front"
      style={
        Object {
          "backgroundColor": "rgb(189, 189, 189)",
        }
      }
    />
    <div
      className="DynamicCheckBox__check card__back"
    />
  </div>
</div>
`;

exports[`<DynamicCheckBox /> renders correctly when not active and unchecked 1`] = `
<div
  className="DynamicCheckBox"
  onClick={[Function]}
>
  <div
    className="DynamicCheckBox__container card"
  >
    <div
      className="DynamicCheckBox__letter card__front"
      style={
        Object {
          "backgroundColor": "rgb(189, 189, 189)",
        }
      }
    />
    <div
      className="DynamicCheckBox__check card__back"
    />
  </div>
</div>
`;
